File: packages/python/alephonenull/v3/core/q_evaluator.py
"""Dataset-level evaluation for the V3 Q score.

Q is a continuous detector score. It still requires labeled harmful and benign
fixtures to measure discrimination and choose an operating threshold.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import isfinite
from typing import List, Optional, Sequence


@dataclass(frozen=True)
class QEvaluationSample:
    expected_unsafe: bool
    Q: float


@dataclass(frozen=True)
class QOperatingPoint:
    threshold: float
    true_positives: int
    false_positives: int
    true_negatives: int
    false_negatives: int
    sensitivity: Optional[float]
    specificity: Optional[float]
    false_positive_rate: Optional[float]
    precision: Optional[float]
    negative_predictive_value: Optional[float]
    f1: Optional[float]
    balanced_accuracy: Optional[float]


@dataclass(frozen=True)
class QEvaluationReport:
    sample_count: int
    unsafe_count: int
    safe_count: int
    mean_unsafe_q: Optional[float]
    mean_safe_q: Optional[float]
    mean_separation: Optional[float]
    auroc: Optional[float]
    operating_point: QOperatingPoint


def _assert_unit_interval(value: float, name: str) -> None:
    if not isfinite(value) or value < 0 or value > 1:
        raise ValueError(f"{name} must be a finite number in [0, 1]")


def _divide(numerator: int, denominator: int) -> Optional[float]:
    return None if denominator == 0 else numerator / denominator


def _mean(values: Sequence[float]) -> Optional[float]:
    return None if not values else sum(values) / len(values)


def _auroc(samples: Sequence[QEvaluationSample]) -> Optional[float]:
    unsafe_count = sum(1 for sample in samples if sample.expected_unsafe)
    safe_count = len(samples) - unsafe_count
    if unsafe_count == 0 or safe_count == 0:
        return None

    sorted_samples = sorted(samples, key=lambda sample: sample.Q)
    unsafe_rank_sum = 0.0
    index = 0
    while index < len(sorted_samples):
        end = index + 1
        while end < len(sorted_samples) and sorted_samples[end].Q == sorted_samples[index].Q:
            end += 1

        average_rank = ((index + 1) + end) / 2
        unsafe_rank_sum += sum(
            average_rank
            for sample in sorted_samples[index:end]
            if sample.expected_unsafe
        )
        index = end

    u = unsafe_rank_sum - unsafe_count * (unsafe_count + 1) / 2
    return u / (unsafe_count * safe_count)


class QEvaluator:
    @staticmethod
    def operating_point(
        samples: Sequence[QEvaluationSample],
        threshold: float = 0.5,
    ) -> QOperatingPoint:
        _assert_unit_interval(threshold, "threshold")
        for sample in samples:
            _assert_unit_interval(sample.Q, "sample.Q")

        true_positives = false_positives = true_negatives = false_negatives = 0
        for sample in samples:
            predicted_unsafe = sample.Q >= threshold
            if sample.expected_unsafe and predicted_unsafe:
                true_positives += 1
            elif not sample.expected_unsafe and predicted_unsafe:
                false_positives += 1
            elif not sample.expected_unsafe:
                true_negatives += 1
            else:
                false_negatives += 1

        sensitivity = _divide(true_positives, true_positives + false_negatives)
        specificity = _divide(true_negatives, true_negatives + false_positives)
        precision = _divide(true_positives, true_positives + false_positives)
        negative_predictive_value = _divide(
            true_negatives, true_negatives + false_negatives
        )
        f1 = _divide(
            2 * true_positives,
            2 * true_positives + false_positives + false_negatives,
        )
        balanced_accuracy = (
            None
            if sensitivity is None or specificity is None
            else (sensitivity + specificity) / 2
        )

        return QOperatingPoint(
            threshold=threshold,
            true_positives=true_positives,
            false_positives=false_positives,
            true_negatives=true_negatives,
            false_negatives=false_negatives,
            sensitivity=sensitivity,
            specificity=specificity,
            false_positive_rate=None if specificity is None else 1 - specificity,
            precision=precision,
            negative_predictive_value=negative_predictive_value,
            f1=f1,
            balanced_accuracy=balanced_accuracy,
        )

    @classmethod
    def evaluate(
        cls,
        samples: Sequence[QEvaluationSample],
        threshold: float = 0.5,
    ) -> QEvaluationReport:
        operating_point = cls.operating_point(samples, threshold)
        unsafe_q: List[float] = [sample.Q for sample in samples if sample.expected_unsafe]
        safe_q: List[float] = [sample.Q for sample in samples if not sample.expected_unsafe]
        mean_unsafe_q = _mean(unsafe_q)
        mean_safe_q = _mean(safe_q)

        return QEvaluationReport(
            sample_count=len(samples),
            unsafe_count=len(unsafe_q),
            safe_count=len(safe_q),
            mean_unsafe_q=mean_unsafe_q,
            mean_safe_q=mean_safe_q,
            mean_separation=(
                None
                if mean_unsafe_q is None or mean_safe_q is None
                else mean_unsafe_q - mean_safe_q
            ),
            auroc=_auroc(samples),
            operating_point=operating_point,
        )

File: packages/npm/src/v3/core/q-evaluator.ts
/**
 * Dataset-level evaluation for the V3 Q score.
 *
 * Q is a continuous detector score. It does not replace label-based
 * evaluation: a versioned harmful/benign dataset is still required to measure
 * discrimination and to choose an operating threshold.
 */

export interface QEvaluationSample {
  /** Ground-truth label supplied by the evaluation dataset. */
  expectedUnsafe: boolean;
  /** Q score returned by the V3 engine. */
  Q: number;
}

export interface QOperatingPoint {
  threshold: number;
  truePositives: number;
  falsePositives: number;
  trueNegatives: number;
  falseNegatives: number;
  sensitivity: number | null;
  specificity: number | null;
  falsePositiveRate: number | null;
  precision: number | null;
  negativePredictiveValue: number | null;
  f1: number | null;
  balancedAccuracy: number | null;
}

export interface QEvaluationReport {
  sampleCount: number;
  unsafeCount: number;
  safeCount: number;
  meanUnsafeQ: number | null;
  meanSafeQ: number | null;
  /** Mean unsafe Q minus mean safe Q. Only defined when both cohorts exist. */
  meanSeparation: number | null;
  /**
   * Threshold-independent ranking quality. Equivalent to the probability that
   * a randomly selected unsafe sample receives a higher Q than a safe sample,
   * with ties worth one half. Only defined when both cohorts exist.
   */
  auroc: number | null;
  operatingPoint: QOperatingPoint;
}

function assertUnitInterval(value: number, name: string): void {
  if (!Number.isFinite(value) || value < 0 || value > 1) {
    throw new RangeError(`${name} must be a finite number in [0, 1]`);
  }
}

function divide(numerator: number, denominator: number): number | null {
  return denominator === 0 ? null : numerator / denominator;
}

function mean(values: number[]): number | null {
  if (values.length === 0) return null;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function average(values: Array<number | null>): number | null {
  const defined = values.filter((value): value is number => value !== null);
  return defined.length === values.length ? mean(defined) : null;
}

/** Mann-Whitney rank statistic with average ranks for tied Q values. */
function calculateAuRoc(samples: QEvaluationSample[]): number | null {
  const unsafeCount = samples.filter(sample => sample.expectedUnsafe).length;
  const safeCount = samples.length - unsafeCount;
  if (unsafeCount === 0 || safeCount === 0) return null;

  const sorted = samples
    .map(sample => ({ ...sample }))
    .sort((a, b) => a.Q - b.Q);

  let unsafeRankSum = 0;
  let index = 0;
  while (index < sorted.length) {
    let end = index + 1;
    while (end < sorted.length && sorted[end]!.Q === sorted[index]!.Q) end++;

    // Ranks are one-based. All tied values receive the average occupied rank.
    const averageRank = ((index + 1) + end) / 2;
    for (let tiedIndex = index; tiedIndex < end; tiedIndex++) {
      if (sorted[tiedIndex]!.expectedUnsafe) unsafeRankSum += averageRank;
    }
    index = end;
  }

  const u = unsafeRankSum - (unsafeCount * (unsafeCount + 1)) / 2;
  return u / (unsafeCount * safeCount);
}

export class QEvaluator {
  static operatingPoint(
    samples: QEvaluationSample[],
    threshold: number = 0.5,
  ): QOperatingPoint {
    assertUnitInterval(threshold, 'threshold');
    for (const sample of samples) assertUnitInterval(sample.Q, 'sample.Q');

    let truePositives = 0;
    let falsePositives = 0;
    let trueNegatives = 0;
    let falseNegatives = 0;

    for (const sample of samples) {
      const predictedUnsafe = sample.Q >= threshold;
      if (sample.expectedUnsafe && predictedUnsafe) truePositives++;
      else if (!sample.expectedUnsafe && predictedUnsafe) falsePositives++;
      else if (!sample.expectedUnsafe) trueNegatives++;
      else falseNegatives++;
    }

    const sensitivity = divide(truePositives, truePositives + falseNegatives);
    const specificity = divide(trueNegatives, trueNegatives + falsePositives);
    const precision = divide(truePositives, truePositives + falsePositives);
    const negativePredictiveValue = divide(
      trueNegatives,
      trueNegatives + falseNegatives,
    );
    const f1 = divide(2 * truePositives, 2 * truePositives + falsePositives + falseNegatives);

    return {
      threshold,
      truePositives,
      falsePositives,
      trueNegatives,
      falseNegatives,
      sensitivity,
      specificity,
      falsePositiveRate: specificity === null ? null : 1 - specificity,
      precision,
      negativePredictiveValue,
      f1,
      balancedAccuracy: average([sensitivity, specificity]),
    };
  }

  static evaluate(
    samples: QEvaluationSample[],
    threshold: number = 0.5,
  ): QEvaluationReport {
    const unsafeQ = samples
      .filter(sample => sample.expectedUnsafe)
      .map(sample => sample.Q);
    const safeQ = samples
      .filter(sample => !sample.expectedUnsafe)
      .map(sample => sample.Q);

    // operatingPoint performs validation for every sample and the threshold.
    const operatingPoint = this.operatingPoint(samples, threshold);
    const meanUnsafeQ = mean(unsafeQ);
    const meanSafeQ = mean(safeQ);

    return {
      sampleCount: samples.length,
      unsafeCount: unsafeQ.length,
      safeCount: safeQ.length,
      meanUnsafeQ,
      meanSafeQ,
      meanSeparation:
        meanUnsafeQ === null || meanSafeQ === null
          ? null
          : meanUnsafeQ - meanSafeQ,
      auroc: calculateAuRoc(samples),
      operatingPoint,
    };
  }
}
